tree/ntuple/v7/test/ntuple_evolution.cxx

```diff
@@ -203,6 +259,135 @@ struct ReorderedMembers {
+      writer->Fill();
+
+      void *ptr = writer->GetModel().GetDefaultEntry().GetPtr<void>("f").get();
+      DeclarePointer("RenamedMember", "ptrRenamedMember", ptr);
+      ProcessLine("ptrRenamedMember->fInt1 = 71;");
+      ProcessLine("ptrRenamedMember->fInt2 = 82;");
+      ProcessLine("ptrRenamedMember->fInt3 = 93;");
+      writer->Fill();
+
+      // Reset / close the writer and flush the file.
+      writer.reset();
+   });
+
+   ASSERT_TRUE(gInterpreter->Declare(R"(
+struct RenamedMember {
+   int fInt1 = 1;
+   int fInt2Renamed = 2;
+   int fInt3 = 3;
+};
+)"));
+
+   auto reader = RNTupleReader::Open("ntpl", fileGuard.GetPath());
+   ASSERT_EQ(2, reader->GetNEntries());
+
+   void *ptr = reader->GetModel().GetDefaultEntry().GetPtr<void>("f").get();
+   DeclarePointer("RenamedMember", "ptrRenamedMember", ptr);
+
+   reader->LoadEntry(0);
+   EXPECT_EVALUATE_EQ("ptrRenamedMember->fInt1", 1);
+   EXPECT_EVALUATE_EQ("ptrRenamedMember->fInt2Renamed", 2);
+   EXPECT_EVALUATE_EQ("ptrRenamedMember->fInt3", 3);
+
+   reader->LoadEntry(1);
+   EXPECT_EVALUATE_EQ("ptrRenamedMember->fInt1", 71);
+   EXPECT_EVALUATE_EQ("ptrRenamedMember->fInt2Renamed", 2);
+   EXPECT_EVALUATE_EQ("ptrRenamedMember->fInt3", 93);
+}
+
+TEST(RNTupleEvolution, AddedBaseClass)
+{
+   FileRaii fileGuard("test_ntuple_evolution_added_base_class.root");
+
+   WriteOldInFork([&] {
+      // The child process writes the file and exits, but the file must be preserved to be read by the parent.
+      fileGuard.PreserveFile();
+
+      ASSERT_TRUE(gInterpreter->Declare(R"(
+struct AddedBaseIntermediate {
+   int fIntermediate = 2;
+};
+struct AddedBaseDerived : public AddedBaseIntermediate {
+   int fDerived = 3;
+};
+)"));
+
+      auto model = RNTupleModel::Create();
+      model->AddField(RFieldBase::Create("f", "AddedBaseDerived").Unwrap());
+
+      auto writer = RNTupleWriter::Recreate(std::move(model), "ntpl", fileGuard.GetPath());
+      writer->Fill();
+
+      void *ptr = writer->GetModel().GetDefaultEntry().GetPtr<void>("f").get();
+      DeclarePointer("AddedBaseDerived", "ptrAddedBaseDerived", ptr);
+      ProcessLine("ptrAddedBaseDerived->fIntermediate = 82;");
+      ProcessLine("ptrAddedBaseDerived->fDerived = 93;");
+      writer->Fill();
+
+      // Reset / close the writer and flush the file.
+      {
+         // TStreamerInfo::Build will report a warning for interpreted classes (but only for base classes).
+         // See also https://github.com/root-project/root/issues/9371
+         ROOT::TestSupport::CheckDiagsRAII diagRAII;
+         diagRAII.optionalDiag(kWarning, "TStreamerInfo::Build", "has no streamer or dictionary",
+                               /*matchFullMessage=*/false);
+         writer.reset();
+      }
+   });
+
+   ASSERT_TRUE(gInterpreter->Declare(R"(
+struct AddedBase {
+   int fBase = 1;
+};
+struct AddedBaseIntermediate : public AddedBase {
+   int fIntermediate = 2;
+};
```

Comment:
Perhaps add a base class to an existing class (prepend). This should check if base classes are matched by type name.

Reply:
You mean a second base class? This currently doesn't work because it changes the field name from `:_0.fBase` to `:_1.fBase`. What works is *append*ing a second base. I added a test for this.

(regarding the type names of base classes, see also above)

Reply:
As discussed offline, we generally forbid the case of mismatching type names of classes for the moment. I added three tests for a renamed member class, a prepended base class, and a renamed base class.